components/widgets/users/avatars/avatars.tsx

```diff
@@ -11,34 +20,88 @@ import SimpleTooltip, {useSynchronizedImmediate} from 'components/widgets/simple
 import Avatar from 'components/widgets/users/avatar';
 
 import './avatars.scss';
+import {imageURLForUser} from 'utils/utils';
+import {RequireOnlyOne} from 'utils/conditional_types';
 
 type StylingKeys = 'size';
 
 type UserProps = ComponentProps<typeof Avatar> & {name: string};
+type AvatarProps = Omit<UserProps, StylingKeys>;
+type Participants = Array<{id: $ID<UserProfile>}>;
 
-export type Props = Pick<UserProps, StylingKeys> & {
-    users: Array<Omit<UserProps, StylingKeys>>;
-    totalUsers?: number;
-    breakAt?: number;
-}
+export type Props =
+    Pick<UserProps, StylingKeys> &
+    RequireOnlyOne<{
+        participants: Participants;
+        users: AvatarProps[];
+    }> & {
+        totalUsers?: number;
+        breakAt?: number;
+    }
 
 const OTHERS_DISPLAY_LIMIT = 99;
 
+function selectUsers(
+    state: GlobalState,
+    users: Props['users'],
+    participants: Props['participants'],
+): null | {users: AvatarProps[]; missingProfiles?: string[]} {
+    if (users) {
```

Comment:
This /seems/ to be a form of memoization? I'm having trouble following this method.

Reply:
It's an API variance short circuit if the more-direct `AvatarProps[]` were provided instead of `Array<{id: UserId}> which requires selection`

Reply:
I opted to remove the API variance and simplify it to just user ids.